lib/internal/async_hooks.js

```diff
@@ -298,71 +297,68 @@ function restoreActiveHooks() {
   active_hooks.tmp_fields = null;
 }
 
-function trackPromise(promise, parent, silent) {
-  const asyncId = getOrSetAsyncId(promise);
+function trackPromise(promise, parent) {
+  if (promise[async_id_symbol]) {
+    return;
+  }
 
+  promise[async_id_symbol] = newAsyncId();
   promise[trigger_async_id_symbol] = parent ? getOrSetAsyncId(parent) :
     getDefaultTriggerAsyncId();
+}
 
-  if (!silent && initHooksExist()) {
-    const triggerId = promise[trigger_async_id_symbol];
-    emitInitScript(asyncId, 'PROMISE', triggerId, promise);
-  }
+function promiseInitHook(promise, parent) {
+  trackPromise(promise, parent);
+  const asyncId = promise[async_id_symbol];
+  const triggerAsyncId = promise[trigger_async_id_symbol];
+  emitInitScript(asyncId, 'PROMISE', triggerAsyncId, promise);
```

Comment:
It has previously been discussed to not expose the actual promise object via `async_hooks` (and eventually `PromiseHooks`), in the hope that one day we could enable escape analysis for promise objects. See https://github.com/nodejs/node/pull/23443 and https://github.com/nodejs/diagnostics/issues/188#issuecomment-386222680

Reply:
That only applies when using PromiseWrap though, which is quite expensive and therefore I've been working towards removing it. Also, AsyncLocalStorage requires the promise, or a distinct resource with the same lifetime as the promise itself, to work properly so there's not currently a way to eliminate that. I'm happy to hear any other suggestions you can think of to handle that.

I've been doing this work, and my prior async_hooks work this year, based on that PromiseHook proposal doc and the later discussions at the summit in Munich. I went for a slightly different compromise path which achieves mostly all the performance wins available in those ideas while avoiding significant breaking changes. This will (hopefully) allow the changes to be backported. We can then consider the remaining breaking parts later. The one part I _do_ still intend to look into from that doc in the short-term is if I can replace `PromiseWrap` gc-tracking with `WeakRef` finalizers instead.